src/pymor/algorithms/lyapunov.py

```diff
@@ -0,0 +1,320 @@
+    prev_shifts
+        A |NumPy array| containing the set of all previously used shift parameters.
+    shift_options
+        The |shift_options| to use (see :func:`lyap_solver_options`).
+
+    Returns
+    -------
+    shifts
+        A |NumPy array| containing a set of stable shift parameters.
+    """
+    u = shift_options['z_columns']
+    L = prev_shifts.size
+    r = len(W)
+    d = L - u
+    if d < 0:
+        u = L
+        d = 0
+    if prev_shifts[L-u].imag < 0:
+        u = u + 1
+
+    Vu = Z[(L - u) * r:L * r]  # last u matrices V added to solution factor Z
+
+    if shift_options['implicit_subspace']:
+        B = np.zeros((u, u))
+        G = np.zeros((u, 1))
+        Ir = np.eye(r)
+        iC = np.where(np.imag(prev_shifts) > 0)[0]  # complex shifts indices (first shift of complex pair)
+        iR = np.where(np.isreal(prev_shifts))[0]  # real shifts indices
+        iC = iC[iC >= d]
+        iR = iR[iR >= d]
+        i = 0
+
+        while i < u:
+            rS = iR[iR < d + i]
+            cS = iC[iC < d + i]
+            rp = prev_shifts[d + i].real
+            cp = prev_shifts[d + i].imag
+            G[i, 0] = np.sqrt(-2 * rp)
+            if cp == 0:
+                B[i, i] = rp
+                if rS.size > 0:
+                    B[i, rS - d] = -2 * np.sqrt(rp*np.real(prev_shifts[rS]))
+                if cS.size > 0:
+                    B[i, cS - d] = -2 * np.sqrt(2*rp*np.real(prev_shifts[cS]))
+                i = i + 1
+            else:
+                sri = np.sqrt(rp**2+cp**2)
+                B[i: i + 2, i: i + 2] = [[2*rp, -sri], [sri, 0]]
+                if rS.size > 0:
+                    B[i, rS - d] = -2 * np.sqrt(2*rp*np.real(prev_shifts[rS]))
+                if cS.size > 0:
+                    B[i, cS - d] = -4 * np.sqrt(rp*np.real(prev_shifts[cS]))
+                i = i + 2
+        B = spla.kron(B, Ir)
+        G = spla.kron(G, Ir)
+
+        s, v = spla.svd(Vu.gramian(), full_matrices=False)[1:3]
+        P = v.T.dot(np.diag(1. / np.sqrt(s)))
+        Q = Vu.to_numpy().T.dot(P)
+
+        E_V = E.apply(Vu).to_numpy().T
+        T = Q.T.dot(E_V)
+        Ap = Q.T.dot(W.to_numpy().T).dot(G.T).dot(P) + T.dot(B.dot(P))
+        Ep = T.dot(P)
+    else:
+        Q = gram_schmidt(Vu, atol=0, rtol=0)
+        Ap = A.apply2(Q, Q)
+        Ep = E.apply2(Q, Q)
+
+    shifts = spla.eigvals(Ap, b=Ep)
+    shifts[np.real(shifts) > 0] = -shifts[np.real(shifts) > 0]
+    if shifts.size == 0:
+        return np.concatenate((prev_shifts, prev_shifts))
+    else:
+        return np.concatenate((prev_shifts, shifts))
+
+
+def _solve_lyap_check_args(A, E, B, trans=False):
+    assert isinstance(A, OperatorInterface) and A.linear
+    assert A.source == A.range
+    assert isinstance(B, OperatorInterface) and B.linear
+    assert not trans and B.range == A.source or trans and B.source == A.source
+    assert E is None or isinstance(E, OperatorInterface) and E.linear and E.source == E.range == A.source
```

Comment:
This function is a copy of `pymor.bindings.scipy._solve_lyap_check_args`. It would make sense now to remove `_solve_lyap_check_args` in `bindings/scipy.py` and add
```python
from pymor.algorithms.lyapunov import _solve_lyap_check_args
```
in `bindings/scipy.py`, `bindings/slycot.py`, and `bindings/pymess.py`.